### PR TITLE
Flag functions with empty body as unimplemented

### PR DESCRIPTION
**Pull request by webthethird**
Flag functions with empty body as unimplemented, unless one or more modifiers are present.

Previously, any function with an empty body, e.g., `function hasArgs(uint, uint) public {}`, would be marked as implemented, with a CFG that contains a single `ENTRY_POINT` node and nothing else. This PR leaves the `ENTRY_POINT` node there, but doesn't set `function.is_implemented = True` unless the function body contains statements (with the exception of empty functions providing access to a modifier, and special functions fallback/receive/constructor).

Some test files expected these empty functions to be marked implemented, so the test contracts have been tweaked to contain one dummy statement, i.e., `uint i = 1;`.

### CODE REVIEW

File: slither/solc_parsing/declarations/function.py [hunk 1]
```diff
@@ -282,7 +282,13 @@ def analyze_content(self) -> None:
...
+                if (
+                    len(body["statements"]) > 0
+                    or len(self._functionNotParsed["modifiers"]) > 0
+                    or self._function.function_type
+                    in [FunctionType.FALLBACK, FunctionType.RECEIVE, FunctionType.CONSTRUCTOR]
+                ):
+                    self._function.is_implemented = True
                 self._parse_cfg(body)
 
             for modifier in self._functionNotParsed["modifiers"]:
```
[the remark this answers is not in the document]
Comment (webthethird) [reply slither/solc_parsing/declarations/function.py:294]: I'm having trouble seeing how the behavior is different. In both the if block (for compact ast) and the else block (for legacy), this still sets `self._function.is_implemented = True` if any of the following are true:
- The body of the function contains one or more statements
- The function has a modifier
- The function is a fallback, receive or constructor

[the remark this answers is not in the document]
Comment (smonicas) [reply slither/solc_parsing/declarations/function.py:294]: You are right. Now i see that the modifier case for the compact ast is together with the other cases and not to be set when iterating the modifiers

**Comment by montyly** [comment]
What is the reason for this change? I think empty function and unimplemented function are two different concepts

**Comment by 0xalpharush** [comment]
@montyly For https://github.com/crytic/slither/pull/1880, we weren't sure that `function hasArgs(uint, uint) public {} returns(uint x)` should generate a RETURN node but it currently has an ENTRY_POINT node. Should we leave `is_implemented` alone and go ahead and generate a RETURN node?

**Comment by montyly** [comment]
yes I think generating a return node would be cleaner. However if we do so, we should add the information that the node was not explicit in Solidity, as this information can be useful for further analyses

**Pull request closed without merging**